In [7]:
import ctypes
import sys
from pympler import asizeof

In [44]:
class DynamicArray:
    def getsize(self):
        import sys 
        try:
            return sys.getsizeof(self._A)
        except:
            return 0
    
    def ToString(self):
        for i in self._A:
            print(i,"")
            
    def getLength(self):
        return len(self._A)
    def __init__(self): 
        self._n=0
        self._capacity=1
        self._A = self._make_array(self._capacity)
        
    def _make_array(self,c):
        return(c*ctypes.py_object)() #malloc a karşılık gelir C de
    
    def append(self,obj):
        if self._n == self._capacity: #Dizi dolu olduğuda
            self._resize(5*self._capacity) # yeniden boyutlandırma yap 
        self._A[self._n] = obj
        self._n +=1
        
        
    def _resize(self,c):
        B = self._make_array(c) #yeni dizi oluştur 
        print("Şu an Amortized Cost işlemi..")
        for k in range(self._n): #A daki elemanları B dizisine aktardı
            B[k] =self._A[k] 
            print("Şu an Move işlemi ...")
           
        self._A = B # A = B 
        self._capacity = c #artan kapasite appenddan önceki kapasitenin 2 katı gelir.
    
    def _len_(self):
        return self._n
   
    
    def _getitem_(self,k):
        if not 0 <= k < self._n:
            raise IndexError ('Invalid Index') 
        return self._A[k]
    def insert(self,k,value):
        if self._n == self._capacity:
            self.resize(2*self_capacity) #yeterli alan olmadığında 
        for j in range(self._n,k,-1): # ilk önce daha sağa kaydırılır 
            self._A[k] = self._A[j-1]
        self._A[k]=value # yeni değeri yerleştir 
        self._n +=1
        
    def remove(self,value):
        #Küçülen bir dinamik dizi düşünülmedi bu versiyonda.
        #Eğer value değeri dizide varsa silinir yoksa Value Error ile sonver
        for k in range(self._n):
            if self._A[k]==value:
                for j in range(k,self._n-1): #boşlukları dolduracak şekilde diğerleri kaydırılır
                    self._A[j] = self._A[j+1]
                self._A[self,n-1]=None  #garbage collectiona yardım eder  (bellek yönetimi)
                self._n-=1 # artık 1 eksildi elemanlar
        return 
        raise ValueError('value not found') #silinecek değer dizide eşleşmedi yok 
        
            
        
    

In [45]:
def get_size(obj,seen=None):
    #Recursive olarak nesnenin boyutunu bulur.
    size = sys.getsizeof(obj)
    if seen is None :
        seen = set()
    obj_id = id(obj)
    if obj_id in seen:
        return 0
    seen.add(obj_id)
    if isinstance(obj,dict):
        size += sum([get_size(v,seen) for v in obj.values()])
        size += sum([get_size(k,seen) for k in obj.keys()])
    elif  hasattr(obj,'_dict_'):
        size += sum([get_size(i,seen) for i in obj])        

    elif hasattr(obj,'_iter_') and not isinstance(obj,(str,bytes,bytearray)):
        size += sum([get_size(i,seen) for i in obj])
    return size 
        

In [56]:
c=DynamicArray()


In [57]:
s_1=sys.getsizeof(c)

s_2=asizeof.asizeof(c)
print("s_1 : {0},s_2 :{1}".format(s_1,s_2))

s_1 : 56,s_2 :640


In [58]:
get_size(c)

56

In [59]:
c.getsize()

128

s_1 : 56  alan var getsize() <br>
s_2 : 640  gerçekte  640 birimlik alan var asizeof()

In [60]:
n=10000
for i in range(n):
    c.append(12)
    #c.append("asdfasdfasdf")
    #c.ToString()
s_1=sys.getsizeof(c)
s_2=asizeof.asizeof(c)
print("s_1 : {0},s_2 :{1}".format(s_1,s_2))

Şu an Amortized Cost işlemi..
Şu an Move işlemi ...
Şu an Amortized Cost işlemi..
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Amortized Cost işlemi..
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Amortized Cost işlemi..
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...


In [61]:
get_size(c)

56

In [62]:
c._A

In [63]:
c.getLength(),c.getsize()

(15625, 128)

In [64]:
for i in range(-100):
    c.append(12)
c.getLength(),c.getsize(),c._len_()

(15625, 128, 10000)

In [65]:
d=DynamicArray()
for i in range(23):
    d.append(i)
    #d.ToString()
     
print("len : {0}".format(d.getLength()))
print("len : {0}".format(d.getsize()))   
   

Şu an Amortized Cost işlemi..
Şu an Move işlemi ...
Şu an Amortized Cost işlemi..
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
Şu an Move işlemi ...
len : 25
len : 128
